src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -507,22 +780,17 @@ private BoundDecisionDag MakeDecisionDag(SyntaxNode syntax, ImmutableArray<Parti
                 case BoundNonNullDecision n1:
                     switch (other)
                     {
-                        case BoundValueDecision v2:
-                            if (v2.Value == ConstantValue.Null)
-                            {
-                                // once v!=null is false, we know v==null is true and do not need to test it
-                                permitsTrue = false;
-                                killsDecisionOnFalseBranch = true;
-                            }
-                            else
-                            {
-                                // Given that v!=null fails, v==K might succeed
-                                permitsFalse = false;
-                            }
+                        case BoundNonNullValueDecision v2:
+                            // Given that v!=null fails, v is null and v==K cannot succeed
+                            falseDecisionPermitsTrueOther = false;
+                            break;
+                        case BoundNullValueDecision v2:
+                            trueDecisionPermitsTrueOther = false; // if v!=null is true, then v==null cannot succeed
```

Comment:
I like the comments, but the English explanation can sometimes make it harder to parse.

Consider using `-->` to represent a logical implication. So, this would be `v != null --> !(v == null)` and the one below would be `!(v != null) --> v == null`

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663) for handling in a future PR

---
In reply to: [176219663](https://github.com/dotnet/roslyn/pull/24059#discussion_r176219663) [](ancestors = 176219663)